nf-core/tools v3.0.0 preparation: update docs for new command nf-core pipelines create (#2742)
nf-core/website: sites/main-site/src/content/blog/2024/tools-3_0_0.md

- nf-validation replaced nf-schema in the pipeline template.
- CI tests now lint with the nf-core tools version matching the template version of the pipeline, to minimise errors in opened PRs with every new tools release.

## 🫡 Deprecations

- The `nf-core licences{:bash}` command is deprecated.

# Avoiding merge conflicts with the new template customisation

If you don't use don't use any of the following template features:

- fastqc
- multiqc
- igenomes
- nf_schema

you can avoid some merge conflicts with a quick update and an intermediate sync:

1. start by checking out the `dev` branch.

```bash
git switch dev
```

1. update the template to the latest version.

```bash
nf-core pipelines sync
```

1. Pull the updated `.nf-core.yml` file from the TEMPLATE branch.

```bash
git checkout TEMPLATE -- .nf-core.yml
```

1. add `fastqc`, `igenomes` or `nf_schema` to skip_features.

```yaml title=".nf-core.yml"
template:
  skip_features:
    - fastqc
    - igenomes
    - nf_schema
```

1. Commit the changes

```bash
git add .nf-core.yml
git commit -m "Skip fastqc, igenomes and nf_schema"
```

> 1. Retrigger the pipeline sync via the [GitHub Actions workflow](https://github.com/nf-core/tools/actions/workflows/sync.yml)

Author: This will run sync to all pipelines, right? Isn't there a way to run it only for the changed one?

Reviewer: added now a manual input to the GHA

> 1. Retrigger the pipeline sync via the [GitHub Actions workflow](https://github.com/nf-core/tools/actions/workflows/sync.yml)

Author: ```suggestion
1. Retrigger the pipeline sync via the [GitHub Actions workflow](https://github.com/nf-core/tools/actions/workflows/sync.yml). Write your pipeline name under the `Pipeline to sync` box.
```

---
title: 'nf-core/tools - 3.0.0'
subtitle: 'What flavour of the nf-core template do you prefer?'
headerImage: 'https://images.unsplash.com/photo-1583258292688-d0213dc5a3a8'
headerImageAlt: 'Yellow and red apples on black plastic crate'
pubDate: 2024-10-03T12:00:00+01:00
authors:
  - 'mirpedrol'
  - 'mashehu'
label:
  - 'tools'
---

> 

Reviewer 2: Need an intro paragraph. Why this is a big release. Why it's major semver. Why it's needed. Why now..

# Highlights

## ⛓️‍💥 Breaking changes

- All pipeline commands now require the `pipelines` prefix.

>   For example, `nf-core lint{:bash}` is now `nf-core pipelines lint{:bash}`.

Reviewer 2: I think that we should list all commands that are changing, from -> to. Maybe in a table?

  This change makes the commands more consistent with `nf-core modules{:bash}` and `nf-core subworkflows{:bash}` commands.
- Some options have been changed for the `nf-core pipelines download{:bash}` command:
  - The `-t` / `--tower` flag has been renamed to `-p` / `--platform`.
  - We renamed the short flags for consistency, to always use the first letter of the second word in the long flag:
    - The `-d` / `--download-configuration` flag has been renamed to `-c` / `--download-configuration`.
    - The `-p` / `--parallel-downloads` flag has been renamed to `-d` / `--parallel-downloads`.

> ## ✨ New features

Reviewer 2: Maybe lead with this? Get people excited before slamming into the work we're asking them to do.

> - Enhanced pipeline template customisation: The template has been divided into features that can be selectively included or excluded.

Reviewer 2: Break up these bullets a bit. Description on sub-bullets.

```suggestion
- Enhanced pipeline template customisation
    - The template has been divided into features that can be selectively included or excluded.
```

Reviewer 2: @mashehu think this one got missed?

Reviewer: not sure I like a bullet list with just on bullet (for the sub-lists)...

Reviewer: changed it anyway

  For example, you can now create a new pipeline without any traces of FastQC.
  You can strip down the pipeline to the bare minimum and add only the tools you need.
  For nf-core pipelines, certain core features (e.g., documentation, CI tests) remain mandatory, but you still have significant customisation flexibility.
- New Text User Interface (TUI) for pipeline creation: A guided interface helps you through the process when running `nf-core pipelines create{:bash}` (don't worry - you can still use the CLI by providing all values as parameters).
- [nf-schema](https://nextflow-io.github.io/nf-schema/latest/) has replaced nf-validation in the pipeline template
- CI tests now use the nf-core tools version matching the pipeline's template version, reducing errors in PRs with each new tools release

## 🫡 Deprecations

- The `nf-core licences{:bash}` command is deprecated.

> # Avoiding Merge Conflicts with Template Customisation

Reviewer 2: ```suggestion
# Tip: Avoiding Merge Conflicts
```

> If you don't use any of these template features:

Reviewer 2: Needs intro paragraph.
```suggestion
The v3 release of nf-core/tools includes the ability to have fine-grained control of which template features to include.
We can use this new functionality to switch off chunks of the template.
Doing so means less code will update in the template, and fewer merge conflicts in files that you don't care about.

So - if you don't use any of these template features:
```

- fastqc
- multiqc
- igenomes
- nf_schema

you can minimize merge conflicts with a quick update and intermediate sync:

> 1. Start by checking out the `dev` branch:
> 
> ```bash
> git switch dev
> ```

Reviewer 2: I think these will work better with indentation:

````suggestion
1. Start by checking out the `dev` branch:

    ```bash
    git switch dev
    ```
````

> 1. Update the template to the latest version:

Reviewer 2: Numbering is off on the website render. I guess because of the indentation thing.

Reviewer: fixed

# Important Template Updates

## The `check_max()` Function Has Been Removed

The `check_max()` function has been replaced by core Nextflow functionality called [`resourceLimits`](https://www.nextflow.io/docs/latest/reference/process.html#resourcelimits).

The `resourceLimits` are specified in the `nextflow.config` file. You can remove all references to `check_max()` and its associated parameters (`max_cpus`, `max_memory`, and `max_time`).
For more information, see the [Nextflow documentation](https://www.nextflow.io/docs/latest/reference/process.html#resourcelimits).

## The nf-validation Plugin Has Been Replaced by nf-schema

The `nf-validation` plugin is deprecated in favour of the new `nf-schema` plugin.

This plugin uses a new JSON schema draft (2020-12), requiring changes to the `nextflow_schema.json` and `assets/schema_input.json` files. Follow the [migration guide](https://nextflow-io.github.io/nf-schema/2.0/migration_guide/) for required changes.

The following validation parameters have been removed from `nextflow.config` and `nextflow_schema.json`:

- validationFailUnrecognisedParams
- validationLenientMode
- validationSchemaIgnoreParams
- validationShowHiddenParams

> - validate_params

Reviewer 2: Code backticks please 🙏🏻

Instead, use the `validation` scope for `nf-schema` options.

:::note

> The plugins definition and `validation` scope have been moved after the `manifest` scope to allow access to `manifest` variables for help message customisation.

Reviewer 2: ```suggestion
The plugins definition and `validation` scope have been moved further down `nextflow.config` and are now found after the `manifest` scope. This allows access to `manifest` variables to help improve message customisation.
```

:::

The `UTILS_NFVALIDATION_PLUGIN` subworkflow has been replaced by `UTILS_NFSCHEMA_PLUGIN`, changing how the input samplesheet is read. See [the documentation](https://nextflow-io.github.io/nf-schema/2.0/migration_guide/#__tabbed_2_2) for details on using the new `samplesheetToList()` function.

> # Removing for-loops and try/catch Blocks from nextflow.config
> 
> To prepare for upcoming Nextflow changes, code in config files has been simplified:

Reviewer 2: ```suggestion
# Removing scripting from `nextflow.config`

To prepare for upcoming Nextflow syntax depreciations, for loops and try/catch blocks have been removed from the template config files:
```

-
- Other code contributors include:
-
- - [Antonia Schuster](https://github.com/AntoniaSchuster)
- - [Alexander Ramos](https://github.com/alxndrdiaz)
- - [Carson Miller](https://github.com/CarsonJM)
- - [Daniel Lundin](https://github.com/erikrikarddaniel)
- - [Danielle Callan](https://github.com/d-callan)
- - [Gregory Sprenger](https://github.com/gregorysprenger)
- - [Jim Downie](https://github.com/prototaxites)
- - [Phil Palmer](https://github.com/PhilPalmer)
- - [@willros](https://github.com/willros)
-
- Long read processing was inspired by [caspargross/HybridAssembly](https://github.com/caspargross/HybridAssembly) written by Caspar Gross [@caspargross](https://github.com/caspargross)
- nf-core/mag was written by [Hadrien Gourlé](https://hadriengourle.com) at [SLU](https://slu.se), [Daniel Straub](https://github.com/d4straub) and [Sabrina Krakau](https://github.com/skrakau) at the [Quantitative Biology Center (QBiC)](http://qbic.life).
=======
+ nf-core/mag was originally written by Hadrien Gourlé, Daniel Straub, Sabrina Krakau, James A. Fellows Yates, Maxime Borry.
>>>>>>> TEMPLATE
```

> ### Resolution

Reviewer 2: I think ditch these subheadings - makes it harder to see delineation between different sections. Also makes the ToC very long.

Content in each section is short so I'm not sure that they're needed.

Reviewer: I link the linkabilty of them, but agree on the ToC messiness. fixed that

-    time   = { check_max( 4.h  * task.attempt, 'time'   ) }
=======
+    // TODO nf-core: Check the defaults for all processes
+    cpus   = { 1      * task.attempt }
+    memory = { 6.GB   * task.attempt }
+    time   = { 4.h    * task.attempt }
>>>>>>> TEMPLATE
```

### Resolution

Double-check the values and accept the incoming changes if correct.

## `.nf-core.yml`

There might be conflicts due to changed order, renamed and new fields, especially in the `template` section.

### Resolution

Double-check the changes for duplicates and accept the incoming changes.

The new file should follow this structure:

```yaml title=".nf-core.yml"
bump_version: null
lint: null
nf_core_version: 3.0.0
org_path: null
template:
  author: Author Name
  description: The description of the pipeline
  force: false
  is_nfcore: true
  name: pipelinename
  org: nf-core
  outdir: .
  skip_features: []
  version: 3.0.0
update: null
```

See the [API docs](/docs/nf-core-tools/api_reference/dev/api/utils#pydantic-modelpythonnf_coreutilsnfcoretemplateconfigpython) for a more detailed description of each field and the allowed input values.

## `nextflow.conifg`

> Several merge conflicts to do changes described in [Important Template Updates](http://localhost:4321/blog/2024/tools-3_0_0#important-template-updates).

Author: ```suggestion
Several merge conflicts due to changes described in [Important Template Updates](http://localhost:4321/blog/2024/tools-3_0_0#important-template-updates).
```

-    memory = { check_max( 6.GB * task.attempt, 'memory' ) }
-    time   = { check_max( 4.h  * task.attempt, 'time'   ) }
=======
+    // TODO nf-core: Check the defaults for all processes
+    cpus   = { 1      * task.attempt }
+    memory = { 6.GB   * task.attempt }
+    time   = { 4.h    * task.attempt }
>>>>>>> TEMPLATE
```

#### Resolution

Double-check the values and accept the incoming changes if correct.

## `.nf-core.yml`

There might be conflicts due to changed order, renamed and new fields, especially in the `template` section.

#### Resolution

Double-check the changes for duplicates and accept the incoming changes.

The new file should follow this structure:

```yaml title=".nf-core.yml"
bump_version: null
lint: null
nf_core_version: 3.0.0
org_path: null
template:
  author: Author Name
  description: The description of the pipeline
  force: false
  is_nfcore: true
  name: pipelinename
  org: nf-core
  outdir: .
  skip_features: []
  version: 3.0.0
update: null
```

See the [API docs](/docs/nf-core-tools/api_reference/dev/api/utils#pydantic-modelpythonnf_coreutilsnfcoretemplateconfigpython) for a more detailed description of each field and the allowed input values.

> ## `nextflow.conifg`

Reviewer 2: ```suggestion
## `nextflow.config`
```

## `.nf-core.yml`

There might be conflicts due to changed order, renamed and new fields, especially in the `template` section.

#### Resolution

Double-check the changes for duplicates and accept the incoming changes.

The new file should follow this structure:

```yaml title=".nf-core.yml"
bump_version: null
lint: null
nf_core_version: 3.0.0
org_path: null
template:
  author: Author Name
  description: The description of the pipeline
  force: false
  is_nfcore: true
  name: pipelinename
  org: nf-core
  outdir: .
  skip_features: []
  version: 3.0.0
update: null
```

See the [API docs](/docs/nf-core-tools/api_reference/dev/api/utils#pydantic-modelpythonnf_coreutilsnfcoretemplateconfigpython) for a more detailed description of each field and the allowed input values.

## `nextflow.conifg`

Several merge conflicts due to changes described in [Important Template Updates](http://localhost:4321/blog/2024/tools-3_0_0#important-template-updates).

#### Resolution

Double-check in the parameters section, that no pipeline-specific parameter would be removed by the incoming changes. In general, you can accept the incoming changes for this file.

> ## `subworkflows/local/utils_nfcore_$PIPELINE_NAME_pipeline/main.nf`
> 
> The switch to `nf-schema` might cause conflicts in the logic of reading in a samplesheet.

Reviewer 2: ```suggestion
## Pipeline utils subworkflow

In `subworkflows/local/utils_nfcore_$PIPELINE_NAME_pipeline/main.nf` the switch to `nf-schema` might cause conflicts in the logic of reading in a samplesheet.
```

- [nf-schema](https://nextflow-io.github.io/nf-schema/latest/) has replaced nf-validation in the pipeline template
- CI tests now use the nf-core tools version matching the pipeline's template version, reducing errors in PRs with each new tools release

## ✨ New features

- Enhanced pipeline template customisation
  The template has been divided into features that can be selectively included or excluded.
  For example, you can now create a new pipeline without any traces of FastQC.
  You can strip down the pipeline to the bare minimum and add only the tools you need.
  For nf-core pipelines, certain core features (e.g., documentation, CI tests) remain mandatory, but you still have significant customisation flexibility.
- New Text User Interface (TUI) for pipeline creation
  A guided interface helps you through the process when running `nf-core pipelines create{:bash}` (don't worry - you can still use the CLI by providing all values as parameters).
- [nf-schema](https://nextflow-io.github.io/nf-schema/latest/) has replaced nf-validation in the pipeline template
- CI tests now use the nf-core tools version matching the pipeline's template version

>   This will reduce errors in opened PRs with each new tools release

Author: @mashehu I think we both did the same change 😄

Reviewer: oops, I removed one of them
